lib/aliquot/payment.rb

```diff
@@ -84,47 +82,153 @@ def protocol_version
     end
 
     def valid_protocol_version?
-      protocol_version == 'ECv1'
+      SUPPORTED_PROTOCOL_VERSIONS.include?(protocol_version)
     end
 
-    ##
-    # Check if the token is expired, according to the messageExpiration included
-    # in the token.
-    def expired?
-      @message[:messageExpiration].to_f / 1000.0 <= Time.now.to_f
+    def validate_intermediate_key
+      # Valid JSON as it has been checked by Token Validator.
+      intermediate_key = JSON.parse(@token[:intermediateSigningKey][:signedKey])
+
+      validator = Aliquot::Validator::SignedKeyValidator.new(intermediate_key)
+      validator.validate
+
+      validator.output
+    end
+
+    def intermediate_key_expired?
+      cur_millis = (Time.now.to_f * 1000).round
+      @intermediate_key[:keyExpiration].to_i < cur_millis
     end
 
-    def valid_signature?
-      signed_string = ['Google', @merchant_id, protocol_version, @token[:signedMessage]].map do |str|
+    def check_shared_secret
+      begin
+        decoded = Base64.strict_decode64(@shared_secret)
+      rescue
+        raise InvalidSharedSecretError, 'shared_secret must be base64'
+      end
+
+      raise InvalidSharedSecretError, 'shared_secret must be 32 bytes when base64 decoded' unless decoded.length == 32
+    end
+
+    def valiate_signature
+      signed_string_message = ['Google', @merchant_id, protocol_version, @token[:signedMessage]].map do |str|
         [str.length].pack('V') + str
       end.join
+      message_signature = Base64.strict_decode64(@token[:signature])
```

Comment:
And, can we even assume that it's strict?

A strategy could be to attempt `strict_decode64`, then `rescue` Base64 errors, log it and `decode64`. Then we'll notice if it's strict or not in practice 🙂

Reply:
This is caught by the validation, as spec'd here:
https://github.com/clearhaus/aliquot/blob/576de21f46606412d829f306ff8223f5acac5ade/spec/validator_spec.rb#L16-L27

So when we get this far it's not a problem.

It really should be strict, I can't imagine Google responding with anything but and i see no reason for technical partners to change the data.
But then again, who knows :)

Reply:
I see your point, however, partners should not be able to trigger an "uncaught" exception at will, IMO. I am not saying that it the case here, though. If it is, the sentence applies :slightly_smiling_face:

Reply:
The validator lines are the following:
https://github.com/clearhaus/aliquot/blob/576de21f46606412d829f306ff8223f5acac5ade/lib/aliquot/validator.rb#L91-L92

Reply:
Imagine that the user of this gem doesn't use the validation (because they have their own validations?) ... dunno how one would usually handle a thing like this.